doc/manual/typing.md

```diff
@@ -0,0 +1,629 @@
+
+### Using untyped code inside typed code
+
+In the other direction, we face a different issue. How can we use untyped values
+inside typed code? Let us try:
+
+```
+let x = 1 in
+(1 + x : Num)
+```
+
+Although `x` shouldn't be statically typed, this code is accepted. Hmm. What
+about:
+
+```
+let x = 0 + 1 in
+(1 + x : Num)
+```
+
+Now, it doesn't, as we would initially expect. In all generality, you can use an
+identifier defined in untyped code from within a typed block, as it is the case
+here. Thus, the typechecker needs to assign a type to every identifier, even the
+ones defined outside of statically typed block. But it also tries to respect the
+intention of the programmer.  If one doesn't use annotations, then the code
+shouldn't be typechecked, whatever the reason is (for performance, because it
+can't be well-typed, etc.): if you want `x` to be typed, you should annotate it.
+
+The typechecker still tries its best not to be too stupid. It is obvious in the
+first case that `1` is of type `Num`. This information is cheap to gather. When
+encountering a binding outside of a typed block, the typechecker determines the
+*apparent type* of the definition. The rationale is that determining the
+apparent type shouldn't recurse arbitrarily inside the expression or do anything
+non-trivial.  Typically, our previous `0 + 1` is a compound expression with a
+primitive operator, so the apparent type is just `Dyn`. For now, the typechecker
+determines an apparent type that is not `Dyn` only for literals (numbers,
+strings, booleans), variables, and annotated expressions.
+
+Otherwise, the typechecker fallbacks to `Dyn`. In the future, it could also
+infer `Dyn -> Dyn` for functions, `{_: Dyn}` for records, and so on. As of now
+(INSERT VERSION HERE), it doesn't.
+
+We could add a type annotation to `x`. But sometimes we don't want to, or we
+can't. Maybe we know that the expression correctly evaluates to a number but the
+typechecker rejects it because it isn't able to see it. In any of these
+situations, we can trade a type annotation for a contract application:
+
+Example:
+```nickel
+let x | Num = if true then 0 else "a" in
+(1 + x : Num)
+```
+
+Here, `x` is clearly always a number, but it is not well-typed (the `then` and
+`else` branches of an `if` must have the same type). Nonetheless, this program
+is accepted! Because we inserted a contract application, the typechecker can be
+sure that if `x` is not a number, the program will fail early with a detailed
+contract error. Thus, if we reach `1 + x`, at this point `x` is necessarily a
+number and won't cause any type mismatch. In a way, the contract application
+acts like a type cast, but whose verification is delayed to run-time.
+
+Dually to a static type annotation, a contract annotation *turns the typechecker
+off again*. This is illustrated by the following variation being accepted:
+
+```nickel
+(1 + ((if true then 0 else "a" | Num)) : Num
+```
```

Comment:
I would start with something like this, to be honest. Then expand to `let x | Num = …` then expand to obvious literals.